sklearn/decomposition/_fastica.py

```diff
@@ -108,7 +108,8 @@ def _ica_par(X, tol, g, fun_args, max_iter, w_init):
         W1 = _sym_decorrelation(np.dot(gwtx, X.T) / p_ - g_wtx[:, np.newaxis] * W)
         del gwtx, g_wtx
         # builtin max, abs are faster than numpy counter parts.
-        lim = max(abs(abs(np.diag(np.dot(W1, W.T))) - 1))
+        # memory improvements and fast execution with np.einsum than np.dot.
+        lim = max(abs(abs(np.einsum("ij,ij->i", W1, W)) - 1))
```

Comment:
I also think the comment above regarding built-in function can be moved after the 2-lines comment here (I can't add it to this suggestion via GitHub UI).
```suggestion
        # np.einsum allows having the lowest memory footprint.
        # It is faster than chaining np.dot with np.diag.
        lim = max(abs(abs(np.einsum("ij,ij->i", W1, W)) - 1))
```